src/cargo/util/command_prelude.rs

```diff
@@ -986,6 +989,49 @@ pub fn root_manifest(manifest_path: Option<&Path>, gctx: &GlobalContext) -> Carg
     }
 }
 
+pub fn get_targets_from_rustup() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustup")
+        .arg("target")
+        .arg("list")
+        .arg("--installed")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
+}
+
+pub fn get_target_from_rustc() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustc")
+        .arg("--print")
+        .arg("target-list")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
+}
+
+pub fn get_target_triple() -> Vec<clap_complete::dynamic::CompletionCandidate> {
+    let mut candidates = Vec::new();
+    if let Ok(targets) = get_targets_from_rustup() {
+        for target in targets {
+            candidates.push(clap_complete::dynamic::CompletionCandidate::new(target));
+        }
+    } else if let Ok(targets) = get_target_from_rustc() {
+        for target in targets {
+            candidates.push(clap_complete::dynamic::CompletionCandidate::new(target));
+        }
+    }
+    candidates
+}
```

Comment:
We should choose our behavior with [`is_rustup`](https://docs.rs/cargo/latest/cargo/util/fn.is_rustup.html)